### PR TITLE
Render user content as Markdown with raw fallback toggle

### PR DESCRIPTION
**Pull request by cboos**
## Summary

User messages have always been escaped into `<pre>` blocks, which is safe but visually coarse — markdown formatting in the source (`**bold**`, `# headings`, backticks) showed up as literal asterisks and pounds. This branch renders user text through the same mistune pipeline as assistant text while preserving the raw view as a one-click fallback.

**Rendering safety**

- `render_user_markdown()` uses `escape=True` so raw `<script>` or `<iframe>` tags are escaped rather than passed through. Assistant-side rendering keeps `escape=False` (by design, since assistants emit trusted structured HTML).
- Every rendered output is validated by `is_well_formed_html()` (stdlib `HTMLParser` tracking a tag stack). XHTML-style self-closing tags (`<br />`, `<hr />`, ``) are recognised as atomic — mistune emits these with `hard_wrap=True`.
- If the output is ill-formed (unclosed tag, mismatched nesting), the message falls back to a bare `<pre>` with no toggle. Well-formed output emits both `.user-md` and `.user-raw` inside a `.user-content` wrapper.

**Toggle UX**

- Per-message: the `.user-view-toggle` button flips one message's wrapper by setting `data-user-view="raw"` or `"md"`.
- Global: the floating `#toggleUserView` button flips `body.show-raw-user`, which cascades to every wrapper that has *no* explicit `data-user-view` attribute. A user who explicitly chose md on a specific message keeps that choice even if global raw is on.
- Default view is Markdown. Default state ships with **no** `data-user-view` attribute so the global toggle actually matches default messages (an earlier iteration baked `data-user-view="md"` in, which broke the global toggle — fixed in 659bce8).
- Global preference persists via `localStorage["claude-code-log:user-view"]`.

**Validation**

- 955 tests / 7 skipped across unit, snapshot, and Playwright browser suites.
- Fixture audit: 0/102 user messages rejected by `is_well_formed_html` (down from 27/102 before the XHTML self-close fix in 3c6fe86).
- pyright, ty, and ruff all clean on the branch tip.

## Test plan

- [x] Unit tests: `uv run pytest -n auto -m \"not (tui or browser)\" -v`
- [x] Browser tests: `uv run pytest -n auto -m browser -v` (5 Playwright scenarios covering default load, global toggle, per-message toggle, explicit-md precedence, and localStorage persistence across reload)
- [x] Type checks: `uv run pyright` and `uv run ty check`
- [x] Lint: `uv run ruff check`
- [x] Snapshot tests: `uv run pytest -n auto test/test_snapshot_html.py`
- [x] Fixture audit: render every user message in `test/test_data/` and confirm 0/102 fall back to the raw-only path

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Per-message and global floating toggles to switch user messages between rendered Markdown and raw text; global choice persists across sessions and shows active state and updated labels.

* **Behavior / Fixes**
  * Well-formed Markdown now displays as rendered HTML alongside a preserved raw view; ill-formed Markdown falls back to raw-only. Per-message explicit choices override the global setting and button labels reflect the next view.

* **Tests**
  * Added unit and browser tests validating dual-view rendering, precedence, persistence, HTML-validation, and edge cases.

### CODE REVIEW

File: claude_code_log/html/templates/transcript.html [hunk 1]
```diff
@@ -206,6 +207,50 @@ <h3>🔍 Search & Filter</h3>
...
+            document.addEventListener('click', function (event) {
+                const btn = event.target.closest('.user-view-toggle');
+                if (!btn) return;
+                const container = btn.closest('.user-content');
+                if (!container) return;
+                const current = container.getAttribute('data-user-view') || 'md';
+                const next = current === 'md' ? 'raw' : 'md';
+                container.setAttribute('data-user-view', next);
+                btn.textContent = next === 'md' ? 'raw' : 'md';
+            });
```
[the remark this answers is not in the document]
Comment (cboos) [reply claude_code_log/html/templates/transcript.html:288]: Yes, I noticed this one as well...

File: claude_code_log/markdown/renderer.py [hunk 2]
```diff
@@ -71,6 +72,52 @@
...
+            if stripped_leading.startswith(fence_delim):
+                fence_delim = None
+            continue
+        fence_match = _FENCE_OPEN_RE.match(line)
+        if fence_match:
+            fence_delim = fence_match.group(1)
+            out.append(line)
+            continue
+        out.append(_HTML_TAG_RE.sub(r"`\1`", line))
+    return "\n".join(out)
```
[the remark this answers is not in the document]
Comment (cboos) [reply claude_code_log/markdown/renderer.py]: That will get hairy really quick... maybe the better way would be to parse and re-render to Markdown with escape on... Not sure if it's possible with mistune, though.

File: claude_code_log/markdown/renderer.py [hunk 3]
```diff
@@ -301,15 +348,30 @@ def title_SessionHeaderMessage(
...
                 parts.append(self._format_image(item))
             elif isinstance(item, TextContent):
                 if item.text.strip():
-                    # Use code fence to protect embedded markdown
-                    parts.append(self._code_fence(item.text))
+                    rendered = render_user_markdown(item.text)
+                    if is_well_formed_html(rendered):
+                        parts.append(_protect_html_tags(item.text))
+                    else:
+                        parts.append(self._code_fence(item.text))
```
[the remark this answers is not in the document]
Comment (cboos) [reply claude_code_log/markdown/renderer.py:489]: That's the paranoid rabbit speaking. The user won't self-inject stuff like this into his prompts, so that he can be surprised when he renders his transcript.

**Comment by cboos** [comment]
This one might be a bit controversial... I mostly always use Markdown in my prompts, so I think it would be nice to see them rendered as such, I made it the default. But I'd also be OK with a more conservative approach (keep the status quo), as there's a global switch md/raw.

**Review by daaain — APPROVED** [review]
This is great! I sometimes use Markdown too and also often paste in random stuff that might break it, but having a toggle as an escape hatch is a good solution
